ast/util.go

```diff
@@ -27,7 +27,12 @@ func IsReadOnly(node Node) bool {
 
 		node.Accept(&checker)
 		return checker.readOnly
-	case *ExplainStmt, *DoStmt:
+	case *ExplainStmt:
+		if !st.Analyze {
+			return true
+		}
+		return IsReadOnly(st.Stmt)
```

Comment:
These 4 lines could simply be `return !st.Analyze || IsReadOnly(st.Stmt)`

Reply:
Nice.
It has been updated.